### ISSUE TITLE
[ATTENTION] Implement Full Grouped Query Attention (GQA) Logic in BitNetAttention::apply_gqa

### ISSUE THREAD

**Issue by EffortlessSteven** [EffortlessMetrics/BitNet-rs#306]
# [ATTENTION] Implement Full Grouped Query Attention (GQA) Logic in `BitNetAttention::apply_gqa`

## Problem Description

The `BitNetAttention::apply_gqa` function in `crates/bitnet-inference/src/layers/attention.rs` currently contains a simplified stub implementation that merely clones the key and value states instead of implementing the complete Grouped Query Attention (GQA) logic. This significantly impacts the model's ability to efficiently handle attention computation when using GQA configurations.

## Environment

- **File**: `crates/bitnet-inference/src/layers/attention.rs`
- **Function**: `BitNetAttention::apply_gqa` (lines 539-547)
- **Codebase**: BitNet.rs neural network inference engine
- **Build Configuration**: Affects both `--features cpu` and `--features gpu`

## Current Implementation (Problematic)

```rust
fn apply_gqa(
    &self,
    key_states: &BitNetTensor,
    value_states: &BitNetTensor,
) -> Result<(BitNetTensor, BitNetTensor)> {
    // For GQA, we repeat the key and value states for each group
    // This is a simplified implementation
    Ok((key_states.clone(), value_states.clone()))
}
```

## Root Cause Analysis

**Technical Investigation Findings:**

1. **Missing GQA Logic**: The function lacks the core GQA mechanism where key/value heads are repeated to match the number of query heads
2. **Performance Impact**: Without proper GQA, the model cannot achieve the memory efficiency benefits that GQA provides
3. **Accuracy Impact**: The current implementation may produce incorrect attention outputs when `num_key_value_heads < num_attention_heads`
4. **Configuration Mismatch**: The `AttentionConfig` supports GQA parameters (`num_key_value_heads`, `num_attention_heads`) but they're not utilized

**Affected Components:**
- Multi-head attention computation
- Memory efficiency in autoregressive generation
- Model accuracy when using GQA configurations
- KV-cache effectiveness

## Impact Assessment

**Severity**: High
**Affected Users**: All users utilizing models with GQA configurations (common in modern LLMs)
**Performance Impact**:
- Memory usage inefficiency (missing GQA memory savings)
- Potential accuracy degradation with GQA-trained models
- Incorrect attention pattern computation

**Business Impact**:
- Reduced model accuracy on GQA models
- Higher memory consumption than necessary
- Potential incompatibility with production GQA models

## Reproduction Steps

1. Build BitNet.rs with attention features: `cargo build --no-default-features --features cpu`
2. Create a model configuration with GQA: `num_attention_heads=32, num_key_value_heads=8`
3. Run inference and observe the attention computation
4. **Expected**: Key/value states repeated 4x (32/8) for each query head group
5. **Actual**: Key/value states unchanged (just cloned)

### Primary Implementation

Implement proper GQA logic that repeats key and value states for each query head group:

```rust
fn apply_gqa(
    &self,
    key_states: &BitNetTensor,
    value_states: &BitNetTensor,
) -> Result<(BitNetTensor, BitNetTensor)> {
    let num_query_heads = self.config.num_attention_heads;
    let num_kv_heads = self.config.num_key_value_heads;

    // Early return if not using GQA
    if num_query_heads == num_kv_heads {
        return Ok((key_states.clone(), value_states.clone()));
    }

    let head_dim = self.config.head_dim;
    let num_kv_groups = num_query_heads / num_kv_heads;

    // Validate that query heads are evenly divisible by kv heads
    if num_query_heads % num_kv_heads != 0 {
        return Err(anyhow::anyhow!(
            "num_attention_heads ({}) must be divisible by num_key_value_heads ({})",
            num_query_heads, num_kv_heads
        ));
    }

    let key_candle = key_states.to_candle()?;
    let value_candle = value_states.to_candle()?;
    let shape = key_candle.shape();

    // Reshape to [batch, seq_len, num_kv_heads, head_dim]
    let key_reshaped = key_candle.reshape(&[
        shape.dims()[0],
        shape.dims()[1],
        num_kv_heads,
        head_dim
    ])?;
    let value_reshaped = value_candle.reshape(&[
        shape.dims()[0],
        shape.dims()[1],
        num_kv_heads,
        head_dim
    ])?;

    // Repeat each kv head num_kv_groups times
    let repeated_key = key_reshaped.repeat_interleave(num_kv_groups, 2)?;
    let repeated_value = value_reshaped.repeat_interleave(num_kv_groups, 2)?;

    // Reshape back to [batch, seq_len, num_query_heads, head_dim]
    let final_key = repeated_key.reshape(&[
        shape.dims()[0],
        shape.dims()[1],
        num_query_heads,
        head_dim
    ])?;
    let final_value = repeated_value.reshape(&[
        shape.dims()[0],
        shape.dims()[1],
        num_query_heads,
        head_dim
    ])?;

    Ok((BitNetTensor::new(final_key), BitNetTensor::new(final_value)))
}
```

### Alternative Approaches

1. **View-based Implementation**: Use tensor views instead of copying data for memory efficiency
2. **CUDA-optimized**: Implement CUDA kernel for GPU-accelerated GQA when `--features gpu`
3. **Lazy Evaluation**: Implement GQA logic during attention computation rather than preprocessing

### Phase 1: Core Implementation (Estimated: 2 days)
- [ ] Implement the full GQA logic with proper tensor reshaping and repetition
- [ ] Add input validation for GQA configuration parameters
- [ ] Add comprehensive error handling for edge cases

### Phase 2: Testing & Validation (Estimated: 1 day)
- [ ] Create unit tests for various GQA configurations (1:1, 4:1, 8:1 ratios)
- [ ] Add integration tests with real model configurations
- [ ] Validate against reference implementations

### Phase 3: Optimization (Estimated: 1 day)
- [ ] Profile memory usage and performance impact
- [ ] Implement GPU-specific optimizations if using `--features gpu`
- [ ] Add benchmarks comparing GQA vs MHA performance

### Phase 4: Documentation (Estimated: 0.5 days)
- [ ] Update inline documentation with GQA algorithm explanation
- [ ] Add examples of GQA usage in model configurations

### Unit Tests
```rust
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_gqa_4_to_1_ratio() {
        // Test 32 query heads with 8 kv heads (4:1 ratio)
    }

    #[test]
    fn test_gqa_no_repetition_when_equal() {
        // Test when num_attention_heads == num_key_value_heads
    }

    #[test]
    fn test_gqa_invalid_configuration() {
        // Test error handling for non-divisible head counts
    }
}
```

### Integration Tests
- Test with real model configurations (Llama-style GQA)
- Cross-validate with C++ reference implementation when available
- Performance benchmarks comparing memory usage before/after

## Risk Assessment

**Potential Complications:**
1. **Memory Layout**: Ensuring tensor memory layout compatibility with downstream attention computation
2. **Performance Regression**: GQA implementation might introduce computational overhead
3. **CUDA Compatibility**: GPU tensor operations may require additional CUDA kernel development

**Mitigation Strategies:**
- Extensive testing with various tensor shapes and configurations
- Performance profiling before and after implementation
- Gradual rollout with feature flags if needed

## Acceptance Criteria

**Functional Requirements:**
- [ ] GQA correctly repeats key/value states according to `num_query_heads / num_key_value_heads` ratio
- [ ] Input validation prevents invalid GQA configurations
- [ ] Error handling provides clear messages for configuration issues
- [ ] Memory usage is optimized (no unnecessary copying when not using GQA)

**Performance Requirements:**
- [ ] GQA implementation adds <5% computational overhead compared to standard MHA
- [ ] Memory usage reduction achieved when using GQA configurations
- [ ] GPU implementation (if applicable) maintains tensor core utilization

**Quality Requirements:**
- [ ] All unit tests pass with >95% code coverage for the function
- [ ] Integration tests validate correctness with real model configurations
- [ ] Cross-validation tests pass when C++ reference is available
- [ ] Documentation includes clear examples and algorithm explanation

## Related Issues

This issue is part of the BitNet.rs attention mechanism improvements. Related issues include:
- Attention mask handling improvements (#TBD - from `simulation-bitnet-attention-compute-attention.md`)
- KV-cache optimizations (#TBD - from various `stub-kv-cache-*.md` files)
- RoPE CUDA optimization (#TBD - from `stub-rotary-embedding-apply-rope-cuda.md`)
- GPU memory management (#TBD - from `github-issue-gpu-memory-management.md`)

## Definition of Done

- [ ] Full GQA implementation replaces stub code
- [ ] All test cases pass including edge cases
- [ ] Performance benchmarks show expected memory savings
- [ ] Code review completed and approved
- [ ] Documentation updated with GQA algorithm details
- [ ] Integration tests validate with production model configurations
